src/sync.rs

```diff
@@ -0,0 +1,121 @@
+#[cfg(not(loom))]
+mod sync_impl {
+    #[cfg(not(feature = "portable-atomic"))]
+    pub(crate) use core::sync::atomic;
+    #[cfg(feature = "portable-atomic")]
+    pub(crate) use portable_atomic as atomic;
+
+    #[cfg(feature = "std")]
+    pub(crate) use parking::{pair, Unparker};
+    #[cfg(feature = "std")]
+    pub(crate) use std::thread;
+
+    pub(crate) trait AtomicWithMut {
+        type Item;
+
+        fn with_mut<R, F: FnOnce(&mut Self::Item) -> R>(&mut self, f: F) -> R;
+    }
+
+    impl AtomicWithMut for atomic::AtomicUsize {
+        type Item = usize;
+
+        fn with_mut<R, F: FnOnce(&mut Self::Item) -> R>(&mut self, f: F) -> R {
+            f(self.get_mut())
+        }
+    }
+
+    impl<T> AtomicWithMut for atomic::AtomicPtr<T> {
+        type Item = *mut T;
+
+        fn with_mut<R, F: FnOnce(&mut Self::Item) -> R>(&mut self, f: F) -> R {
+            f(self.get_mut())
+        }
+    }
+
+    pub(crate) struct UnsafeCell<T>(core::cell::UnsafeCell<T>);
+
+    impl<T> UnsafeCell<T> {
+        pub(crate) const fn new(item: T) -> Self {
+            Self(core::cell::UnsafeCell::new(item))
+        }
+
+        pub(crate) fn with_mut<F, R>(&self, f: F) -> R
+        where
+            F: FnOnce(*mut T) -> R,
+        {
+            f(self.0.get())
+        }
+    }
+}
+
+#[cfg(loom)]
+mod sync_impl {
+    pub(crate) use loom::cell::UnsafeCell;
+    pub(crate) use loom::thread;
+
+    pub(crate) mod atomic {
+        pub(crate) use core::sync::atomic::compiler_fence;
+        pub(crate) use loom::sync::atomic::*;
+    }
+
+    use loom::sync::{Arc, Condvar, Mutex};
+
+    /// Re-implementation of `parking::pair` based on loom.
```

Comment:
Can we upstream this into https://github.com/smol-rs/parking ?

Reply:
Sure! It should be a straightforward reimplementation in terms of Loom primitives.

Reply:
now that it's upstreamed, can we use that version? (although I don't know how it would interact with dependencies...) Does cargo correctly work with git deps which also have versions (e.g. transforms them into crates.io entries on publishing)?